consensus/istanbul/core/core.go

```diff
@@ -515,20 +556,33 @@ func (c *core) startNewRound(round *big.Int) error {
 		Round:    new(big.Int).Set(round),
 	}
 
-	var err error
-	request, roundChangeCertificate, err := c.getPreprepareWithRoundChangeCertificate(round)
-	if err != nil {
-		logger.Error("Unable to produce round change certificate", "err", err, "new_round", round)
-		return nil
-	}
-
 	// Calculate new proposer
 	prevProposer := c.current.Proposer()
 	prevBlock := c.current.Sequence().Uint64() - 1
 	blockAuthor := c.backend.AuthorForBlock(prevBlock)
 	valSet := c.current.ValidatorSet()
 	nextProposer := c.selectProposer(valSet, blockAuthor, newView.Round.Uint64())
 
+	var err error
+	var request *istanbul.Request
+	var roundChangeCertificate istanbul.RoundChangeCertificate
+	var roundChangeCertificateV2 istanbul.RoundChangeCertificateV2
+	if c.address == nextProposer.Address() && propose {
```

Comment:
This is a funny one:
`startNewRound` is called from two different places: `handleRoundChange` and `handleRoundChangeCertificate`. The first occurs when receiving a `RoundChange(V1 or V2)` message, and the second when receiving a `Preprepare(V1 or V2)` message (round >= 1).

So, in the second case, during a preprepare handling, this function is creating a preprepare and a roundchangecertificate with the round change messages that it received in the roundchangecertificate. So it's pointless, but this is how it was written before since it's shared code called from two different sides. 

However, with our V2 change of the RoundChangeCertificate, we may not have the round change messages, but still received the preprepare with a RCC. And the RoundChangeSet stores roundchangeMessages since it needs the full block to decide which is the one that goes with the highest round preparedCertificate. So we don't have the data to create it properly. The solution was to modify completely how the roundChangeSet works, but since this was a change to fix something that shouldn't be done anyways, I just added the propose flag. A greater rework may imply bigger changes (calling two different functions) but I wanted to keep changes at minimum.

Reply:
Ok, makes sense. Could you add a comment for this to a void a future mistake?

Reply:
Done